### PR TITLE
Refactor parsed file field

### PR DESCRIPTION
**Pull request by tobiasdiez**
Inspired by a discussion with @lenhard in #2692. In detail, the following changes were performed:
- Split `FileField` into `FileFieldWriter` and `FileFieldParser`
- Move `getFiles` from `TypedBibEntry` to `BibEntry` (the later already contained a `setFiles` method)
- Add `isOnlineLink` and `findIn` to `ParsedFileField`
- Extract a few methods from `logic.util.io.FileUtil` to `model.util.FileHelper` (idea for a better name?)
- Rename `ParsedFileField` to `LinkedFile`

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)
- [ ] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`?

### CODE REVIEW

File: src/main/java/org/jabref/model/util/FileHelper.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        return getFileExtension(file.getName());
+    }
+
+    /**
+     * Returns the extension of a file name or Optional.empty() if the file does not have one (no "." in name).
+     *
+     * @param fileName
+     * @return The extension (without leading dot), trimmed and in lowercase.
+     */
+    public static Optional<String> getFileExtension(String fileName) {
```
Comment (koppor) [inline src/main/java/org/jabref/model/util/FileHelper.java]: Can you add a JavaDoc comment why you implemented that method on your own instead of using [Files.getFileExtension(fileName);](http://stackoverflow.com/a/15985486/873282)? We have the Guava libraries and should use them 😇

File: src/main/java/org/jabref/model/entry/FileFieldWriter.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        return getStringRepresentation(Collections.singletonList(field));
+    }
+
+    /**
+     * Encodes a two-dimensional String array into a single string, using ':' and
+     * ';' as separators. The characters ':' and ';' are escaped with '\'.
+     * @param values The String array.
+     * @return The encoded String.
+     */
+    public static String encodeStringArray(String[][] values) {
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/FileFieldWriter.java]: Can you add a JavaDoc why this method is public? The other below is private, I miss the symmetry.

File: src/main/java/org/jabref/model/entry/BibEntry.java [hunk 3]
```diff
@@ -847,6 +848,21 @@ private void invalidateFieldCache(String fieldName) {
...
+    /**
+     * Gets a list of linked files.
+     *
+     * @return the list of linked files, is never null but can be empty
+     */
+    public List<LinkedFile> getFiles() {
+        //Extract the path
+        Optional<String> oldValue = getField(FieldName.FILE);
+        if (!oldValue.isPresent()) {
+            return new ArrayList<>();
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/BibEntry.java]: Why not `Collections.emptyList()`?

File: src/main/java/org/jabref/model/entry/LinkedFile.java [hunk 4]
```diff
@@ -1,23 +1,31 @@
...
 import java.util.Objects;
+import java.util.Optional;
 
-public class ParsedFileField {
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.metadata.FileDirectoryPreferences;
+import org.jabref.model.util.FileHelper;
 
-    private static final ParsedFileField NULL_OBJECT = new ParsedFileField("", "", "");
+public class LinkedFile {
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/LinkedFile.java]: Could you add a JavaDoc stating the intension and usage of this class? Also, why no TypedBibEntry is required anymore?

File: src/main/java/org/jabref/model/util/FileHelper.java (hunk 1 above)
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/util/FileHelper.java]: I thik he just moved the method, we previously had it in FileUtils

File: src/main/java/org/jabref/gui/externalfiles/SynchronizeFileField.java [hunk 5]
```diff
@@ -160,8 +153,8 @@ public void run() {
                         boolean deleted = false;
 
                         // Get an absolute path representation:
-                        Optional<File> file = FileUtil.expandFilename(flEntry.getLink(), dirsS);
-                        if ((!file.isPresent()) || !file.get().exists()) {
+                        Optional<Path> file = field.findIn(panel.getBibDatabaseContext(), Globals.prefs.getFileDirectoryPreferences());
+                        if ((file.isPresent()) && Files.exists(file.get())) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/externalfiles/SynchronizeFileField.java]: Why did you change the logic here?` Previously it was ! present

File: src/main/java/org/jabref/model/util/FileHelper.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    public static Optional<Path> expandFilename(final BibDatabaseContext databaseContext, String name,
+                                                        FileDirectoryPreferences fileDirectoryPreferences) {
+        Optional<String> extension = getFileExtension(name);
+        // Find the default directory for this field type, if any:
+        List<String> directories = databaseContext.getFileDirectories(extension.orElse(null), fileDirectoryPreferences);
+        // Include the standard "file" directory:
+        List<String> fileDir = databaseContext.getFileDirectories(fileDirectoryPreferences);
+        // Include the directory of the BIB file:
+        List<String> al = new ArrayList<>();
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/util/FileHelper.java]: A better variable name would be nice

File: src/main/java/org/jabref/model/util/FileHelper.java (hunk 1 above)
Comment (koppor) [inline src/main/java/org/jabref/model/util/FileHelper.java]: We nevertheless could use Guava 😇

Comment (tobiasdiez) [inline src/main/java/org/jabref/model/util/FileHelper.java]: Our code is slightly different (extensions are required to have at length > 1) and uses optional instead of empty strings. Thus I think it is not worth using Guava (maybe we can even remove the dependency on guava at some point).

File: src/main/java/org/jabref/model/entry/FileFieldWriter.java (hunk 2 above)
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/entry/FileFieldWriter.java]: It is public because it is used somewhere else. You are probably right and this method should be private, but I have no idea what "FileListTableModel" is doing...

File: src/main/java/org/jabref/model/entry/BibEntry.java (hunk 3 above)
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/entry/BibEntry.java]: done

File: src/main/java/org/jabref/model/entry/LinkedFile.java (hunk 4 above)
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/entry/LinkedFile.java]: Of course I can.

File: src/main/java/org/jabref/gui/externalfiles/SynchronizeFileField.java (hunk 5 above)
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/externalfiles/SynchronizeFileField.java]: Nice catch! Was probably a copy&paste error on my side. Fixed

File: src/main/java/org/jabref/model/util/FileHelper.java (hunk 6 above)
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/util/FileHelper.java]: Yep
